// matrix-js-sdk/lib/crypto/algorithms/base.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

exports.registerAlgorithm = registerAlgorithm;

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/*
Copyright 2016 OpenMarket Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/**
 * Internal module. Defines the base classes of the encryption implementations
 *
 * @module
 */

/**
 * map of registered encryption algorithm classes. A map from string to {@link
 * module:crypto/algorithms/base.EncryptionAlgorithm|EncryptionAlgorithm} class
 *
 * @type {Object.<string, function(new: module:crypto/algorithms/base.EncryptionAlgorithm)>}
 */
var ENCRYPTION_CLASSES = exports.ENCRYPTION_CLASSES = {};

/**
 * map of registered encryption algorithm classes. Map from string to {@link
 * module:crypto/algorithms/base.DecryptionAlgorithm|DecryptionAlgorithm} class
 *
 * @type {Object.<string, function(new: module:crypto/algorithms/base.DecryptionAlgorithm)>}
 */
var DECRYPTION_CLASSES = exports.DECRYPTION_CLASSES = {};

/**
 * base type for encryption implementations
 *
 * @alias module:crypto/algorithms/base.EncryptionAlgorithm
 *
 * @param {object} params parameters
 * @param {string} params.userId  The UserID for the local user
 * @param {string} params.deviceId The identifier for this device.
 * @param {module:crypto} params.crypto crypto core
 * @param {module:crypto/OlmDevice} params.olmDevice olm.js wrapper
 * @param {module:base-apis~MatrixBaseApis} baseApis base matrix api interface
 * @param {string} params.roomId  The ID of the room we will be sending to
 * @param {object} params.config  The body of the m.room.encryption event
 */

var EncryptionAlgorithm = function () {
    function EncryptionAlgorithm(params) {
        _classCallCheck(this, EncryptionAlgorithm);

        this._userId = params.userId;
        this._deviceId = params.deviceId;
        this._crypto = params.crypto;
        this._olmDevice = params.olmDevice;
        this._baseApis = params.baseApis;
        this._roomId = params.roomId;
    }

    /**
     * Encrypt a message event
     *
     * @method module:crypto/algorithms/base.EncryptionAlgorithm.encryptMessage
     * @abstract
     *
     * @param {module:models/room} room
     * @param {string} eventType
     * @param {object} plaintext event content
     *
     * @return {module:client.Promise} Promise which resolves to the new event body
     */

    /**
     * Called when the membership of a member of the room changes.
     *
     * @param {module:models/event.MatrixEvent} event  event causing the change
     * @param {module:models/room-member} member  user whose membership changed
     * @param {string=} oldMembership  previous membership
     * @public
     */


    _createClass(EncryptionAlgorithm, [{
        key: 'onRoomMembership',
        value: function onRoomMembership(event, member, oldMembership) {}
    }]);

    return EncryptionAlgorithm;
}();

exports.EncryptionAlgorithm = EncryptionAlgorithm; // https://github.com/jsdoc3/jsdoc/issues/1272

/**
 * base type for decryption implementations
 *
 * @alias module:crypto/algorithms/base.DecryptionAlgorithm
 *
 * @param {object} params parameters
 * @param {string} params.userId  The UserID for the local user
 * @param {module:crypto} params.crypto crypto core
 * @param {module:crypto/OlmDevice} params.olmDevice olm.js wrapper
 * @param {string=} params.roomId The ID of the room we will be receiving
 *     from. Null for to-device events.
 */

var DecryptionAlgorithm = function () {
    function DecryptionAlgorithm(params) {
        _classCallCheck(this, DecryptionAlgorithm);

        this._userId = params.userId;
        this._crypto = params.crypto;
        this._olmDevice = params.olmDevice;
        this._roomId = params.roomId;
    }

    /**
     * Decrypt an event
     *
     * @method module:crypto/algorithms/base.DecryptionAlgorithm#decryptEvent
     * @abstract
     *
     * @param {object} event raw event
     *
     * @return {null} if the event referred to an unknown megolm session
     * @return {module:crypto.DecryptionResult} decryption result
     *
     * @throws {module:crypto/algorithms/base.DecryptionError} if there is a
     *   problem decrypting the event
     */

    /**
     * Handle a key event
     *
     * @method module:crypto/algorithms/base.DecryptionAlgorithm#onRoomKeyEvent
     *
     * @param {module:models/event.MatrixEvent} params event key event
     */


    _createClass(DecryptionAlgorithm, [{
        key: 'onRoomKeyEvent',
        value: function onRoomKeyEvent(params) {}
        // ignore by default


        /**
         * Import a room key
         *
         * @param {module:crypto/OlmDevice.MegolmSessionData} session
         */

    }, {
        key: 'importRoomKey',
        value: function importRoomKey(session) {
            // ignore by default
        }
    }]);

    return DecryptionAlgorithm;
}();

exports.DecryptionAlgorithm = DecryptionAlgorithm; // https://github.com/jsdoc3/jsdoc/issues/1272

/**
 * Exception thrown when decryption fails
 *
 * @alias module:crypto/algorithms/base.DecryptionError
 * @param {string} msg user-visible message describing the problem
 *
 * @param {Object=} details key/value pairs reported in the logs but not shown
 *   to the user.
 *
 * @extends Error
 */

var DecryptionError = function (_Error) {
    _inherits(DecryptionError, _Error);

    function DecryptionError(msg, details) {
        _classCallCheck(this, DecryptionError);

        var _this = _possibleConstructorReturn(this, (DecryptionError.__proto__ || Object.getPrototypeOf(DecryptionError)).call(this, msg));

        _this.name = 'DecryptionError';
        _this.details = details;
        return _this;
    }

    /**
     * override the string used when logging
     *
     * @returns {String}
     */


    _createClass(DecryptionError, [{
        key: 'toString',
        value: function toString() {
            var _this2 = this;

            var result = this.name + '[msg: ' + this.message;

            if (this.details) {
                result += ', ' + Object.keys(this.details).map(function (k) {
                    return k + ': ' + _this2.details[k];
                }).join(', ');
            }

            result += ']';

            return result;
        }
    }]);

    return DecryptionError;
}(Error);

exports.DecryptionError = DecryptionError; // https://github.com/jsdoc3/jsdoc/issues/1272

/**
 * Exception thrown specifically when we want to warn the user to consider
 * the security of their conversation before continuing
 *
 * @param {string} msg message describing the problem
 * @param {Object} devices userId -> {deviceId -> object}
 *      set of unknown devices per user we're warning about
 * @extends Error
 */

var UnknownDeviceError = exports.UnknownDeviceError = function (_Error2) {
    _inherits(UnknownDeviceError, _Error2);

    function UnknownDeviceError(msg, devices) {
        _classCallCheck(this, UnknownDeviceError);

        var _this3 = _possibleConstructorReturn(this, (UnknownDeviceError.__proto__ || Object.getPrototypeOf(UnknownDeviceError)).call(this, msg));

        _this3.name = "UnknownDeviceError";
        _this3.devices = devices;
        return _this3;
    }

    return UnknownDeviceError;
}(Error);

/**
 * Registers an encryption/decryption class for a particular algorithm
 *
 * @param {string} algorithm algorithm tag to register for
 *
 * @param {class} encryptor {@link
 *     module:crypto/algorithms/base.EncryptionAlgorithm|EncryptionAlgorithm}
 *     implementation
 *
 * @param {class} decryptor {@link
 *     module:crypto/algorithms/base.DecryptionAlgorithm|DecryptionAlgorithm}
 *     implementation
 */


function registerAlgorithm(algorithm, encryptor, decryptor) {
    ENCRYPTION_CLASSES[algorithm] = encryptor;
    DECRYPTION_CLASSES[algorithm] = decryptor;
}
